core/src/nodes/handled_node.rs

```diff
@@ -22,28 +22,26 @@ use muxing::StreamMuxer;
 use nodes::node::{NodeEvent, NodeStream, Substream};
 use futures::{prelude::*, stream::Fuse};
 use std::io::Error as IoError;
-use Multiaddr;
 
 /// Handler for the substreams of a node.
-///
-/// > Note: When implementing the various methods, don't forget that you have to register the
-/// > task that was the latest to poll and notify it.
 // TODO: right now it is possible for a node handler to be built, then shut down right after if we
 //       realize we dialed the wrong peer for example ; this could be surprising and should either
 //       be documented or changed (favouring the "documented" right now)
-pub trait NodeHandler<TSubstream> {
+pub trait NodeHandler {
```

Comment:
Why not keeping `NodeHandler` generic over the substream type?

Reply:
If `TSubstream` is generic, that means you can implement the `NodeHandler` trait multiple times on the same struct with different types for `TSubstream`.

However the implementation of `NodeHandler` may need to store fields that depend on `TSubstream` and therefore it doesn't make sense to allow multiple trait implementations.

That's the theoretical explanation, in practice I've encountered several inference issues with `TSubstream` being generic.